nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -0,0 +1,524 @@
+        # only the last stages of the pipeline return losses
+        if losses_reduced_per_micro_batch:
+            # NOTE: assume that the returned values are already gathered across the DP workers
+            rewards_chosen = torch.cat([item["out_actual"] for item in losses_reduced_per_micro_batch])
+            rewards_rejected = torch.cat([item["out_generated"] for item in losses_reduced_per_micro_batch])
+
+            rewards_all = torch.cat((rewards_chosen, rewards_rejected))
+            rewards_chosen_mean = rewards_chosen.mean()
+            rewards_rejected_mean = rewards_rejected.mean()
+            rewards_all_mean = rewards_all.mean()
+            rewards_all_std = rewards_all.std()
+
+            # average loss across micro batches
+            loss_tensors_list = [loss_reduced["avg"] for loss_reduced in losses_reduced_per_micro_batch]
+            loss_tensor = torch.concat(loss_tensors_list)
+            loss_mean = loss_tensor.mean()
+            acc_tensors_list = [loss_reduced["acc"] for loss_reduced in losses_reduced_per_micro_batch]
+
+            if len(acc_tensors_list) == 1:
+                acc_tensor = acc_tensors_list[0]
+            elif len(acc_tensors_list) > 1:
+                acc_tensor = torch.concat(acc_tensors_list)
+            acc_mean = acc_tensor.mean()
+        else:
+
+            loss_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            acc_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            rewards_chosen_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_rejected_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_std = torch.tensor(0.0, device=torch.cuda.current_device())
+
+        # we can only log on one rank if it is rank zero so we broadcast from last rank
+        torch.distributed.broadcast(loss_mean, get_last_rank())
+        torch.distributed.broadcast(acc_mean, get_last_rank())
+
+        torch.distributed.broadcast(rewards_chosen_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_rejected_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_std, get_last_rank())
+
+        metrics = {
+            "loss": loss_mean,
+            "acc": acc_mean,
+            "rewards_actual_mean": rewards_chosen_mean,
+            "rewards_generated_mean": rewards_rejected_mean,
+            "rewards_all_mean": rewards_all_mean,
+            "rewards_all_std": rewards_all_std,
+        }
+
+        # move to CPU
+        metrics = {k: v.item() for k, v in metrics.items()}
+
+        return loss_mean.item(), metrics
+
+    def get_loss_and_metrics_vanilla_sft(self, batch, forward_only):
+        """Take a data_iter which is an interator over the microbatches
+            and return loss as well as metrics
```

Comment:
This takes a `batch` as input, not a `data_iter`

Reply:
Fixed the comment